src/main/java/org/prebid/server/bidder/OpenrtbBidder.java

```diff
@@ -0,0 +1,207 @@
+        final Map<T, List<Imp>> extToImps = new LinkedHashMap<>();
+        for (Imp imp : bidRequest.getImp()) {
+            try {
+                validateImp(imp);
+                final T impExt = parseAndValidateImpExt(imp, getTypeReference());
+                extToImps.putIfAbsent(impExt, new ArrayList<>());
+                extToImps.get(impExt).add(modifyImp(imp, impExt));
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+        final List<Imp> modifiedImps = extToImps.values().stream()
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+        if (modifiedImps.isEmpty()) {
+            errors.add(BidderError.badInput("No valid impression in the bid request"));
+            return Result.of(Collections.emptyList(), errors);
+        }
+        final MultiMap headers = makeHeaders(bidRequest);
+
+        return Result.of(createHttpRequests(extToImps, modifiedImps, bidRequest, headers, errors), errors);
+    }
+
+    // hook for bidder-specific request validation if any is required
+    protected void validateRequest(BidRequest bidRequest) throws PreBidException {
+    }
+
+    //hook for bidder-specific imp validation if any is required
+    protected void validateImp(Imp imp) {
+    }
+
+    private T parseAndValidateImpExt(Imp imp, TypeReference typeReference) {
+        final T impExt;
+        try {
+            impExt = Json.mapper.<ExtPrebid<?, T>>convertValue(imp.getExt(), typeReference).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+
+        validateImpExt(impExt, imp);
+
+        return impExt;
+    }
+
+    //hook for bidder-specific imp.ext validation if any is required
+    protected void validateImpExt(T impExt, Imp imp) {
+    }
+
+    // hook for impression changes.
+    // To be defined by client as most bidders apply some changes to outgoing impressions.
+    // If no changes are made or to be made later on - just return an incoming impression.
+    protected abstract Imp modifyImp(Imp imp, T impExt);
+
+    // hook for request headers changes.
+    // By default - returns basic headers
+    protected MultiMap makeHeaders(BidRequest bidRequest) {
+        return BidderUtil.headers();
+    }
+
+    // hook for bidder-specific http requests creation logic.
+    // By default - a one-to-one implementation (modifies incoming request and forwards it to bidder)
+    // Can be modified to handle one-to-many scenarios (creating multiple requests from single incoming request)
+    protected List<HttpRequest<BidRequest>> createHttpRequests(Map<T, List<Imp>> extToImps, List<Imp> modifiedImps,
+                                                               BidRequest bidRequest, MultiMap headers,
+                                                               List<BidderError> errors) {
```

Comment:
During my research I've found that some bidders could have an additional error-proof logic even at this stage. Although I agree that it is better for force bidders to do the validation sooner rather than later, yet some of the existing ones could require an errors list to be accessible in case of any exceptional situations.

Reply:
Solution would be to move validation logic to `validateRequest`, `validateImp` or `validateImpExt` if possible when transforming those bidders or if it's not possible - give up on transforming them. Anyways, there shouldn't be any validations beyond `validate...` methods.